### PR TITLE
Use floating point inputs in the mouse emulation code (SDL 3 preparation)

### PR DESCRIPTION
**Pull request by FeralChild64**
# Description

- reworked internal mouse APIs (used for mouse wheel and seamless mouse) to accept floating point parameters, as provided by SDL 3 (SDL 2 only provides integer ones)
- some cleanups and code unification

## Related issues

None, but related to PR https://github.com/dosbox-staging/dosbox-staging/pull/4041

# Manual testing

- checked few games (_Civilization_, _Settlers II_, _Master of Orion II_) and applications (_Norton Commander_, _DOS Navigator_) still work with the emulated mouse driver (checked both seamless and captured mouse)
- checked serial (COM1), PS/2 (both register-level and BIOS-level access) with the software mentioned above, using _CtMouse_, as described on the Wiki https://github.com/dosbox-staging/dosbox-staging/wiki/Mouse-emulation-tests
- checked _VirtualBox_ and _VMware_ mouse interfaces with _Norton Commander_ and _DOS Navigator_ using the _VBADOS_ mouse driver (see the Wiki page above)
- for all the mouse interfaces/drivers providing wheel support, checked it using _Mpxplay_ 1.66
- checked that _Windows 3.11 for Workgroups_ still runs with the stock mouse driver

# Checklist

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] provided the release notes draft (for significant user-facing changes).
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: src/hardware/input/mouseif_dos_driver.cpp [hunk 1]
```diff
@@ -1280,8 +1278,17 @@ void MOUSEDOS_NotifyMoved(const float x_rel, const float y_rel,
...
+		if (std::lround(pending.x_abs / Epsilon) != std::lround(x_abs / Epsilon) ||
+		    std::lround(pending.y_abs / Epsilon) != std::lround(y_abs / Epsilon)) {
 			event_needed = true;
+		}
+		// TODO: Consider introducing some kind of sensitivity to avoid
+		// unnecessary events, for example calculated using
+		// 'state.maxpos_*' and 'mouse_shared.resolution_*' values.
+		// Problem: when the mouse is moved really fast and it leaves
+		// the window, the guest cursor is sometimes left like 30 pixels
+		// from the window's edge; we need a mitigation mechanism here
```
Comment (johnnovak) [inline src/hardware/input/mouseif_dos_driver.cpp:1291]: I'd say that's fine. Solution: don't move it very fast if you don't want the pointer to stop before the edge? 😄 

Btw, I don't see why it matters; maybe for seamless mode?

File: src/hardware/input/mouse_common.cpp [hunk 2]
```diff
@@ -90,16 +91,58 @@ uint8_t MOUSE_GetDelayFromRateHz(const uint16_t rate_hz)
...
 {
+	constexpr float Max = 2048.0f;
 	// Enforce sane upper limit of relative mouse movement
-	return std::clamp(rel, -2048.0f, 2048.0f);
+	return std::clamp(rel, -Max, Max);
+}
+
+float MOUSE_ClampWheelMovement(const float rel)
+{
+	constexpr float Max = 127.0f; // so that it always fits into int8_t
```
Comment (johnnovak) [inline src/hardware/input/mouse_common.cpp:101]: Minor: no long end of the line comments, please.

File: src/hardware/input/mouseif_dos_driver.cpp [hunk 3]
```diff
@@ -96,12 +96,12 @@ static MouseButtons12S pending_button_state = 0;
...
-static float pos_x             = 0.0f;
-static float pos_y             = 0.0f;
-static int8_t counter_w        = 0; // wheel counter
+static float pos_x = 0.0f;
+static float pos_y = 0.0f;
+static int8_t counter_w = 0; // wheel counter
 
-static bool use_relative    = true; // true = ignore absolute mouse position, use relative
-static bool is_input_raw    = true; // true = no host mouse acceleration pre-applied
+static bool use_relative = true; // true = ignore absolute mouse position
```
Comment (johnnovak) [inline src/hardware/input/mouseif_dos_driver.cpp:103]: Can you move the comments to the preceding lines?

**Comment by johnnovak** [comment]
Nice job @FeralChild64, I appreciate you've started breaking things up into separate commits 💯 

Make sure to test with stretched `viewport` settings too, e.g.:

```
[render]
aspect = stretch
viewport = relative 112% 173%
integer_scaling = off
```

**Review by interloper98 — APPROVED** [review]
Just a very minor comment. Don't know much about this code, but everything looks reasonable.

File: src/hardware/input/mouse_common.cpp (hunk 2 above)
Comment (FeralChild64) [reply src/hardware/input/mouse_common.cpp]: OK, changed.

File: src/hardware/input/mouseif_dos_driver.cpp (hunk 3 above)
Comment (FeralChild64) [reply src/hardware/input/mouseif_dos_driver.cpp]: OK, changed.

File: src/hardware/input/mouse_common.cpp [hunk 4]
```diff
@@ -88,16 +91,58 @@ uint8_t MOUSE_GetDelayFromRateHz(const uint16_t rate_hz)
...
+
+int8_t MOUSE_ConsumeInt8(float& delta, const bool skip_delta_update)
+{
+	if (!MOUSE_HasAccumulatedInt(delta)) {
+		return 0;
+	}
+
+	const auto consumed = std::round(delta);
+	if (!skip_delta_update) {
+		delta -= static_cast<float>(consumed);
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/hardware/input/mouse_common.cpp]: You are right - removed unnecessary cast.

**Comment by johnnovak** [comment]
> > Make sure to test with stretched `viewport` settings too, e.g.:
> 
> Looks it is working correctly too. But it was much easier to test with shrunk viewport :)

Yeah, the setting I posted enlarges the viewport to fix the aspect ratio of lazy Hercules conversions that mostly just reuse the 320x200 art (e.g., Prince of Persia).

**Comment by FeralChild64** [comment]
@johnnovak Do you still want to test/review something? Or can we merge it?

**Pull request merged**